libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt

```diff
@@ -0,0 +1,302 @@
+        writeJoined(codeList, "\n\n", indent)
+    }
+
+    override fun GeneratedCodeVisitor.visit() {
+        for (code in codeList) {
+            visit(code)
+        }
+    }
+}
+
+/**
+ * References a member that can be used as a type
+ */
+interface IdentifiableMember {
+    val name: String
+    fun imports(): Set<String> = emptySet()
+
+    companion object {
+        private data class KClassIdentifiable(val kClass: KClass<*>) : IdentifiableMember {
+            override val name: String = kClass.simpleName ?: kClass.qualifiedName!!
+            override fun imports() = setOfNotNull(kClass.qualifiedName)
+        }
+
+        private data class NamedIdentifiable(override val name: String) : IdentifiableMember
+
+        operator fun invoke(kClass: KClass<*>): IdentifiableMember = KClassIdentifiable(kClass)
+        operator fun invoke(name: String): IdentifiableMember = NamedIdentifiable(name)
+    }
+}
+
+/**
+ * References an interface that can be implemented
+ */
+interface GeneratedInterface : IdentifiableMember
+
+private data class CastedInterfaceMember(private val member: IdentifiableMember) : GeneratedInterface, IdentifiableMember by member
+
+fun IdentifiableMember.asInterface(): GeneratedInterface = CastedInterfaceMember(this)
+
+/**
+ * Generated code that is a type that can be used as an Argument
+ */
+interface IdentifiableCode : GeneratedCode, IdentifiableMember
+
+/**
+ * Builder for a provided type whose JVM Name can be set
+ */
+interface JVMNameableBuilder {
+    fun jvmName(name: String)
+}
+
+/**
+ * Builder for a provided type. It can implement an interface and have some code inside.
+ */
+interface GeneratedTypeBuilder : IdentifiableMember {
+    /**
+     * Implement a generated interface
+     */
+    fun implement(generated: GeneratedInterface)
+
+    /**
+     * Add code to the namespace of the Type
+     */
+    fun inner(code: GeneratedCode)
+}
+
+/**
+ * Builder for a provided data class. It can implement interfaces, have code inside and have a number of properties
+ */
+interface DataClassBuilder : JVMNameableBuilder, GeneratedTypeBuilder {
+    fun property(name: String, type: IdentifiableMember, default: GeneratedCode? = null)
+}
+
+/**
+ * Builder for a provided object. It can implement interfaces, have code inside and have a number of properties.
+ */
+interface ObjectBuilder : JVMNameableBuilder, GeneratedTypeBuilder
+
+/**
+ * Add multiple instances of generated code inside of the type
+ */
+fun GeneratedTypeBuilder.inner(init: GeneratedCode.Builder.() -> Unit) {
+    inner(GeneratedCode.composite(init))
+}
+
+/**
+ * Implement the interface.
+ */
+fun GeneratedTypeBuilder.implement(kClass: KClass<*>) {
+    require(kClass.java.isInterface) { "Provided type is not an interface" }
+    implement(IdentifiableMember(kClass).asInterface())
+}
+
+/**
+ * Add a companion object to the generated type
+ */
+fun GeneratedTypeBuilder.companionObject(name: String? = null, init: ObjectBuilder.() -> Unit) {
```

Comment:
It would be nice to have tests that use all functions from the DSL, otherwise, they may rot over time and we may discover it too late.

Reply:
Added a bunch of tests. It should be just shy of 70% coverage now